src/aktualizr_primary/secondary.cc

```diff
@@ -141,20 +147,53 @@ class SecondaryWaiter {
   std::unordered_set<std::string> secondaries_to_wait_for_;
 };
 
-static Secondaries createIPSecondaries(const IPSecondariesConfig& config) {
+static Secondaries createIPSecondaries(const IPSecondariesConfig& config, bool provision, Aktualizr& aktualizr) {
   Secondaries result;
-  SecondaryWaiter sec_waiter{config.secondaries_wait_port, config.secondaries_timeout_s, result};
+  if (provision) {
+    SecondaryWaiter sec_waiter{config.secondaries_wait_port, config.secondaries_timeout_s, result};
 
-  for (auto& ip_sec_cfg : config.secondaries_cfg) {
-    auto secondary = Uptane::IpUptaneSecondary::connectAndCreate(ip_sec_cfg.ip, ip_sec_cfg.port);
-    if (secondary) {
-      result.push_back(secondary);
-    } else {
-      sec_waiter.addSecondary(ip_sec_cfg.ip, ip_sec_cfg.port);
+    for (auto& ip_sec_cfg : config.secondaries_cfg) {
+      auto secondary = Uptane::IpUptaneSecondary::connectAndCreate(ip_sec_cfg.ip, ip_sec_cfg.port);
+      if (secondary) {
+        result.push_back(secondary);
+      } else {
+        sec_waiter.addSecondary(ip_sec_cfg.ip, ip_sec_cfg.port);
+      }
+    }
+
+    sec_waiter.wait();
+  } else {
+    auto secondaries_info = aktualizr.GetRegisteredSecondaries();
+
+    // for backward compatibility, we take all secondaries if their number
+    // matches, without matching on their type
+    if (secondaries_info.size() != config.secondaries_cfg.size()) {
+      secondaries_info.erase(
+          std::remove_if(secondaries_info.begin(), secondaries_info.end(),
+                         [](const SecondaryInfo& info) -> bool { return info.type != IPSecondariesConfig::Type; }));
+    }
+
+    if (secondaries_info.size() != config.secondaries_cfg.size()) {
+      std::stringstream err_ss;
+      err_ss << "Expected " << config.secondaries_cfg.size() << " IP secondaries, but device was provisioned with "
```

Comment:
In this case we have to make sure that device is not provisioned if one or more of the secondaries listed in the conf is not available during the provisioning phase.

Reply:
Indeed! This should be covered by what we added in #1491.